src/Limiters/limiter.jl

```diff
@@ -0,0 +1,130 @@
+"""
+    quasimonotone_limiter!(ρq, ρ, min_ρq, max_ρq; rtol)
+
+Arguments:
+- `ρq`: tracer density Field, where `q` denotes tracer concentration per unit mass
+- `ρ: fluid density Field
+- `min_ρq: Array of min(ρq) per element
+- `max_ρq`: Array of max(ρq) per element
+- `rtol`: relative tolerance needed to solve element-wise optimization problem
+
+This limiter is inspired by the one presented in Guba et al [GubaOpt2014](@cite).
+In the reference paper, it is denoted by OP1, and is outlined in eqs. (37)-(40).
+Quasimonotone here is meant to be monotone with respect to the spectral element
+nodal values. This limiter involves solving a constrained optimization problem
+(a weighted least square problem up to a fixed tolerance denoted by `rtol`)
+that is completely local to each element.
+As in HOMME, the implementation idea here is the following: we need to find a grid
+field which is closest to the initial field (in terms of weighted sum), but satisfies
+the min/max constraints. So, first we find values that do not satisfy constraints
+and bring these values to a closest constraint. This way we introduce some mass
+change (`mass_change`), which we then redistribute so that the
+l2 error is smallest. This redistribution might violate constraints; thus, we do
+a few iterations (typically a couple).
+"""
+function quasimonotone_limiter!(
+    ρq::Fields.Field,
+    ρ::Fields.Field,
+    min_ρq,
+    max_ρq;
+    rtol,
+)
+    space = axes(ρ)
+    FT = Spaces.undertype(space)
+    Nq = Spaces.Quadratures.degrees_of_freedom(Spaces.quadrature_style(space))
+    n_elems = length(min_ρq)
+    q = zeros(axes(Fields.slab(ρ, 1)))
+    ρ_wJ = zeros(axes(Fields.slab(ρ, 1)))
+
+    # Traverse elements
+    for e in 1:n_elems
+        ρ_e_slab = Fields.slab(ρ, e)
+        ρq_e_slab = Fields.slab(ρq, e)
+        # Compute ρ's and ρq's masses over an element e
+        ρ_e_mass = sum(ρ_e_slab)
+        ρq_e_mass = sum(ρq_e_slab)
+
+        # This should never happen, but if it does, don't limit
+        if ρ_e_mass <= 0
+            continue
+        end
+
+        # Relax constraints to ensure limiter has a solution:
+        # This is only needed if running with the SSP CFL>1 or
+        # due to roundoff errors.
+        if ρq_e_mass < min_ρq[e] * ρ_e_mass
+            min_ρq[e] = ρq_e_mass / ρ_e_mass
+        end
+
+        if ρq_e_mass > max_ρq[e] * ρ_e_mass
+            max_ρq[e] = ρq_e_mass / ρ_e_mass
+        end
+
+        local_geometry_slab = Fields.slab(space.local_geometry, e)
+
+        q = ρq_e_slab ./ ρ_e_slab
```

Comment:
Is this meant to be
```julia
        q .= ρq_e_slab ./ ρ_e_slab
```
?

Reply:
At the moment, that would lead to the following error:
```
Broacasted spaces are the same ClimaCore.Spaces type but not the same instance
```

Reply:
You are creating multiple instances of the same space.

Reply:
Charlie's point was that how it is currently written you are creating lots of small intermediate arrays

Reply:
Yeah, I guess we'd need to add a caching mechanism if we wanted to make it less allocating. I'm not sure how big these arrays are, so maybe this isn't a big deal. If you wanted to do that (add a caching mechanism), we could change the interface to something like

```julia
quasimonotone_limiter!(args...) -> apply!(limiter::QuasimonotoneLimiter, args...)
```
Where we can cache things so that `apply!` is non-allocating by re-using the preallocated fields (e.g., `q`) in `QuasimonotoneLimiter`.
```julia
struct QuasimonotoneLimiter{T}
        q::T
end
```

Reply:
Thanks Charlie. This `q` is only as big as the number of quadrature points on one 2D element, i.e., `Nq*Nq`,  which is 16 for a typical use case with cubic polynomial degree (p=3) Finite Element basis. I would personally leave any further optimization including the caching to subsequent PRs. Especially if we want to add different limiters, it would make more sense.

Reply:
you can workaround this issue by using the backing data layouts, the field slabs need to be integrated but everything else can use slab datalayouts in the computations.   the pre allocated slabs "cache" are at the top of the function but you don't use them